### ISSUE TITLE
EPIC-06 — Perception and action kernels on the 4090, then the Orin

### ISSUE THREAD

**Issue by superposition** [superposition/qualia#6]
### Phase 6 — Perception and action kernels on the 4090, then the Orin (steps 16–18)

**Step 16. Add the three kernels.**
In `crates/cuda/kernels/`, add `belief_couple.cu` (prior-weighted slot update, 1 thread per slot),
`perception_voxel.cu` (lidar points to occupancy logits, 1 thread per voxel, mirroring
`leash-cuda`'s voxel kernel's input shape: flat `[x, y, ...]` coordinates), and `action_score.cu`
(one block per rollout candidate scoring `terminal_goal_distance_m` and
`max_collision_probability` against a candidate's steps). Register all three in the same module list
the existing kernels use and add CPU oracles beside the existing ones in `crates/cuda/src/lib.rs`
following the established oracle pattern.

**Step 17. Prove parity.**
Extend `crates/jepa-model/src/parity.rs`'s pattern with `crates/cuda/tests/parity.rs` covering the
three new kernels: for each, run the CPU oracle and the CUDA path over fixed, empty, maximum-size and
deterministic-random inputs and assert relative error `<= 1e-4` for f32. Command:
`cargo test -p qualia-cuda --features cuda`.

**Step 18. Budget for 8 GB and build for both targets.**
Add `crates/cuda/tests/budget.rs::fits_orin_nano` asserting that the allocation plan returned by a new
`pub fn memory_plan(shape: &CostmapStatsShape) -> u64` is `<= 6 * 1024 * 1024 * 1024` bytes, leaving
2 GB of the Nano's 8 GB for the runtime. Add a `just`-style Make target in the existing `Makefile`:
`cuda-fatbin-87: NVCC=/usr/local/cuda/bin/nvcc cargo build -p qualia-cuda --features cuda --release`
and compile for `sm_87` (Orin) and `sm_89` (4090) by setting `CUDAARCHS=87-real;89-real`. The 4090 is
the development target and `QUALIA_CUDA_SM` (already read by `runners/cuda-service`) selects the
device at runtime.

> **Clean-room decision in effect.** The public repository is a from-scratch rewrite under
> Apache-2.0. No file is copied from any private repository; the private engine is read only for
> its interfaces, and the connectome dataset only for its graph. See `docs/decisions.md`.

## Tickets

- T16 — #31 Three `.cu` kernels + CPU oracles
- T17 — #32 CPU/CUDA parity
- T18 — #33 `memory_plan` ≤ 6 GiB, fatbin targets

## Journal

Entry: [`the-fly-brain-on-the-robot`](https://github.com/superposition/qualia/tree/main/docs/figures/the-fly-brain-on-the-robot) — three editors (accuracy, teaching, style) approved it on one head and `qualia-gates journal --pr 258` reported `journal-gate: OK`.

The slug this epic previously named, `the-fly-in-the-belief-matrices`, belongs to an earlier epic and was not reused.

**Comment by superposition** [comment]
```braid
agent: JournalEpic06Writer
branch: journal/the-fly-brain-on-the-robot
commit: 50ed851
state: review
next: the three editors read PR #258 (accuracy / teaching / style) per docs/journal-review.md; publish after three approves
blocked_on: none
evidence: https://github.com/superposition/qualia/pull/258
```

The journal entry for this epic is drafted: **PR #258**, based on `main`, `Refs #6`. It covers
#235/#236/#237/#238/#239/#240/#241/#244/#247/#250/#251 and D-023–D-027, with three figures drawn by a
committed generator and every number traced to a named source.

**Slug change.** This entry's slug is **`the-fly-brain-on-the-robot`**. The `## Journal` line in this
issue's body currently names `the-fly-in-the-belief-matrices`, which the EPIC-04/05/06 entry already
uses (`docs/figures/the-fly-in-the-belief-matrices/`); the repository README's journal table now
carries the new row against #6. Whoever updates this issue on close should set the `## Journal` line to
`https://superposition.github.io/journal/the-fly-brain-on-the-robot/`. I did not edit the issue body.

Two numbers the entry defends under questioning: **11.840 ms/tick (84.5 ticks/s) on the Orin NX iGPU**
over the 190 MB artifact, and **`valid=1499 candidates=1500`** on `t65-real-01` — each a line of
committed evidence (`docs/evidence/T55/connectome-runner/README.md`;
`docs/evidence/T65/calibration/README.md`).

Gates at `50ed851`: `cargo run --quiet -p qualia-gates -- figures` → `figures: OK (11 entries, 55
figures, 400 KiB budget)`, exit 0; `cargo run --quiet -p qualia-gates -- provenance` → `provenance: OK`,
exit 0; the entry leg (`journal --entry …`) reports `entry OK` and exits 1 until the roles leg runs
with `--pr 258`. Not merged by me.

**Comment by superposition** [comment]
EPIC-06 closes with every ticket under it landed and its journal entry published.

**The wave, in one paragraph.** The released FlyEM male-CNS connectome is imported, committed as an artifact under `assets/brain/connectome-cns/` (CC-BY, digests verified) and stepped as a network — **166,700 neurons, 25,582,938 neuron-level edges, 124,177,617 synapses** — at **11.840 ms/tick → 84.5 ticks/s on the Orin NX iGPU** (182 MB resident, one tick per frame at 60 Hz with ~1.4× headroom), **1.381 ms/tick → 18.5 G synapses/s on the 4090**, with the host loop running 200 ticks off the robot's real camera at 112.7 ticks/s and the board loop 31–32 ticks/s. The brain is drawn inside the operator's own console from that committed artifact with the real firing stream. The robot's own sensors reach MCAP through the leash (a 900 s capture, 83,299,459 B, sha256 `97612acb…`), and three new producers — calibration, pose, and a leash-authority applied-action ledger recording the leash's own acknowledged zero-speed `stop` — took the dataset from **0 of 4501** candidates accepted to **1,499 of 1,500**. The mission broker made the tree's first decision-producing model call (DeepSeek, decision `promote`, envelope accepted by the agent with HTTP 202), and the review that followed found and closed a credential leak in its status surface. The four Python gates became Rust with content-identical parity. The device audit proved leash owns the hardware — both serial ports return `Device or resource busy` — and fixed the stack manifest that had been starting neither the lidar subscriber nor the camera, so nothing had fed the arena from the robot.

**The engineering the wave recorded as failures** (the useful part, per the journal template): a numpy byte-index mask that under-set bits and made the first connectome join 24× too small, caught only because a second implementation disagreed; `serde_json`'s default float parser moving a 17-digit literal by 1 ulp so a written manifest failed **its own digest** (309 of 4,727 adversarial literals parse wrong without `float_roundtrip`); a CPU LIF path aliasing its own spike buffer, masked by a bench that fired nothing; a parity guard that accepted a torn belief on any architecture; the QLSP reader offset by four bytes; and two board-lease violations that produced **D-024**, the lease rule.

**The decisions this wave added**: D-023 (the console is the operator's instrument; one agent instance per host; no windows on the operator's desktop), D-024 (the board is a leased resource), D-026 (committed snapshot images are re-rendered on the rebased tree, never sided), D-027 (the promotion gate stands uncalibrated — the real corpus is 1,499 valid transitions in 1 session / 1 environment / 1 condition against 50k / 12 / 3 / 3, so closing it is operations, not code).

**What it does not establish**, stated in the entry and repeated here so the epic does not read as more than it is: no learning, plasticity or neuromodulation — the only state is membrane potential and a refractory counter; the encoder grid/gain and decoder rule are chosen constants, not fitted; the camera faced a nearly uniform scene; no promotion-passing checkpoint exists yet; the console figure is one host's wgpu render; and the `tcp://` live spike path is unexercised. Plumbing, not behaviour.

**Journal**: `docs/figures/the-fly-brain-on-the-robot/` — entry, two charts, one diagram, `make_figures.py`, `wave-numbers.json`; three editors read it (accuracy, teaching, style), each approval posted as a PR comment, and `qualia-gates journal --pr 258` reported `roles OK` and `journal-gate: OK`.
